### PR TITLE
Add benchmark for NetCDF --> Zarr cloud-optimization

### PR DESCRIPTION
**Pull request by jrbourbeau**
NetCDF datasets being slow/not scaling well has come up a lot. This PR adds a new benchmark that loads the `nex-gddp-cmip6` dataset (https://registry.opendata.aws/nex-gddp-cmip6/) from AWS, which is stored as a bunch of `.nc` files, and converts that dataset to Zarr, a more modern, cloud-optimized format.

This is using `xr.open_mfdataset(..., parallel=True)` which is both common and _really_ slow when opening lots of NetCDF files, which I like because I've seen this with many users in practice. 

One thing I'm not sure about is how representative this benchmark is as is. I don't know if folks do this NetCDF --> Zarr conversion in isolation, or always in conjunction with other "cloud optimizing" steps like rechunking. 

EDIT: Here's a cluster link for the "small" version of this test https://cloud.coiled.io/clusters/594106/account/dask-engineering/information. It takes ~20 minutes and costs ~$0.75

### CODE REVIEW

File: tests/geospatial/test_netcdf_to_zarr.py
Comment (jrbourbeau) [inline tests/geospatial/test_netcdf_to_zarr.py:93]: @maxrjones thanks for pointing to this dataset over in https://github.com/coiled/benchmarks/discussions/1545#discussioncomment-10681943. Does this looks like what you've seen in the wild?

File: .github/workflows/tests.yml
Comment (jrbourbeau) [inline .github/workflows/tests.yml:34]: TODO: Revert before merging

File: tests/geospatial/test_netcdf_to_zarr.py
Comment (jrbourbeau) [reply tests/geospatial/test_netcdf_to_zarr.py]: I've gone ahead and added a rechunking step (from "pancake" to "pencil" chunks), which seems to be pretty common when cloud-optimizing a NetCDF dataset to Zarr

Comment (maxrjones) [reply tests/geospatial/test_netcdf_to_zarr.py]: Yeah, I think it's right that when cloud-optimizing it's usually best to include a rechunking step. But, there's many cases in which people avoid the cloud-optimization step and do subsequent analyses on the data loaded directly from the original NetCDF files.

File: tests/geospatial/test_cloud_optimize.py
Comment (dcherian) [inline tests/geospatial/test_cloud_optimize.py:89]:
```suggestion
            parallel=True, data_vars="minimal", coords="minimal", compat="override",
```

These *may* be needed for decent perf, I haven't looked at the files to be sure

File: tests/geospatial/test_netcdf_to_zarr.py
Comment (jrbourbeau) [reply tests/geospatial/test_netcdf_to_zarr.py]: > Yeah, I think it's right that when cloud-optimizing it's usually best to include a rechunking step.

Awesome, thanks for confirming that's usually the case 👍 

> But, there's many cases in which people avoid the cloud-optimization step and do subsequent analyses on the data loaded directly from the original NetCDF files

I guess in this case, the "subsequent analyses" is just writing to Zarr. Do you think this still captures user pain well? FWIW my experience is using `xr.open_mfdataset(..., parallel=True)` + any other step performs very poorly

File: tests/geospatial/test_cloud_optimize.py
Comment (dcherian) [reply tests/geospatial/test_cloud_optimize.py:89]: ^ yes, they are basically mandatory

Comment (jrbourbeau) [reply tests/geospatial/test_cloud_optimize.py:89]: Cool, I'll take them for a spin and update here. Also, should those be the default in `xarray`?

Comment (dcherian) [reply tests/geospatial/test_cloud_optimize.py:89]: https://github.com/pydata/xarray/issues/8778

File: tests/geospatial/test_cloud_optimize.py
Comment (hendrikmakait) [inline tests/geospatial/test_cloud_optimize.py:68]: Is there a difference here in how we scale with respect to models and variables? I'm wondering if it would make more sense to pick a subset (larger than one) of both models and variables instead.

File: tests/geospatial/test_cloud_optimize.py
Comment (hendrikmakait) [inline tests/geospatial/test_cloud_optimize.py:84]: nit: 
```suggestion
        # TODO: Reduce explicit settings once https://github.com/pydata/xarray/issues/8778 is completed.
        ds = xr.open_mfdataset(
```

File: tests/geospatial/test_cloud_optimize.py
Comment (hendrikmakait) [inline tests/geospatial/test_cloud_optimize.py:10]:
```suggestion
        "workspace": "dask-benchmarks",
```

**Review by hendrikmakait — APPROVED** [review]
Thanks, @jrbourbeau, this is good to go once the workspace has been adjusted. I have one question regarding scaling this benchmark but that's non-blocking.

File: tests/geospatial/test_cloud_optimize.py
Comment (jrbourbeau) [reply tests/geospatial/test_cloud_optimize.py]: 👍

File: tests/geospatial/test_netcdf_to_zarr.py
Comment (maxrjones) [reply tests/geospatial/test_netcdf_to_zarr.py]: > I guess in this case, the "subsequent analyses" is just writing to Zarr. Do you think this still captures user pain well? FWIW my experience is using xr.open_mfdataset(..., parallel=True) + any other step performs very poorly

Yeah, it could make sense to focus first on the simplest operations that perform poorly. You mention that going from pancake to pencil chunks is a common step that performs poorly. That's true and is the motivation for the rechunker library. The worst performance would likely be seen when combining operations that perform optimally on pancake oriented chunks with operations that perform optimally on churro oriented chunks.

Comment (jrbourbeau) [reply tests/geospatial/test_netcdf_to_zarr.py]: > pancake oriented chunks...churro oriented chunks

💯

**Pull request merged**
